docs/spec.md

```diff
@@ -140,240 +43,355 @@ stablehlo.func @main(
-end implementation-defined.
+```ebnf
+Op            ::= [OpOutputs] OpName OpInputs ':' OpSignature
+OpName        ::= '"' 'stablehlo' '.' OpMnemonic '"'
+OpMnemonic    ::= 'abs' | 'add' | ...
+```
 
-Further formalization, e.g. how streaming communication influences execution
-order and what kind of synchronization is introduced by it, is TBD.
+**StableHLO operations** (which are also called **ops**) have a name,
+inputs/outputs and a signature. The name consists of the `stablehlo.` prefix and
+a **mnemonic** which uniquely identifies one of the supported ops. See below for
+a comprehensive list of all supported ops.
+
+```ebnf
+OpInputs        ::= OpInputValues OpInputFuncs OpInputAttrs
+OpInputValues   ::= '(' [OpInputValue {',' OpInputValue}] ')'
+OpInputValue    ::= ValueId
+OpInputFuncs    ::= ['(' OpInputFunc {',' OpInputFunc} ')']
+OpInputAttrs    ::= ['{' OpInputAttr {',' OpInputAttr} '}']
+OpOutputs       ::= [OpOutput {',' OpOutput} '=']
+OpOutput        ::= ValueId
+```
 
-### Collective ops
+Ops consume **inputs** and produce **outputs**. Inputs are categorized into
+input values (computed during execution), input functions (provided
+statically, because in StableHLO functions are not first-class values) and
+input attributes (also provided statically). The kind of inputs and outputs
+consumed and produced by an op depends on its mnemonic. For example, the `add`
+op consumes 2 input values and produces 1 output value. In comparison, the
+`select_and_scatter` op consumes 3 input values, 2 input functions and
+3 input attributes.
+
+```ebnf
+OpInputFunc ::= '{' Unused FuncInputs ':' FuncBody '}'
+Unused      ::= '^' digit {digit}
+              | '^' letter {letter | digit}
+```
 
-There are five collective ops in StableHLO: `all_gather`, `all_reduce`,
-`all_to_all`, `collective_permute` and `reduce_scatter`. All these ops split
-the processes in the StableHLO grid into **StableHLO process groups** and
-execute a joint computation within each process group, independently from other
-process groups.
+**Input functions** (which are also called **anonymous functions**) are very
+similar to named functions except that: 1) they don't have an identifier (hence
+the name "anonymous"), 2) they don't declare output types (output types are
+inferred from the `return` op within the function).
+
+The syntax for input functions includes a currently unused part (see the
+`Unused` production above) which is there for compatibility with MLIR. In MLIR,
+there is a more general concept of "regions" which can have multiple "blocks"
+of ops connected together via jump ops. These blocks have ids which correspond
+to the `Unused` production, so that they can be distinguished from each other.
+StableHLO doesn't have jump ops, so the corresponding part of MLIR syntax is
+unused (but is still there).
+
+```ebnf
+OpInputAttr      ::= OpInputAttrName '=' OpInputAttrValue
+OpInputAttrName  ::= letter {letter | digit}
```

Comment:
Similar to functions, these allow dots as well. However, we leave that area rather unspecced since there isn't a good use case for these. So I think leaving it out is fine.